core/src/test/java/org/apache/iceberg/variants/TestValueArray.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.iceberg.util.RandomUtil;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class TestValueArray {
+  private static final VariantMetadata EMPTY_METADATA = Variants.emptyMetadata();
+  private static final List<VariantValue> ELEMENTS =
+      ImmutableList.of(
+          Variants.of(34), Variants.of("iceberg"), Variants.of(new BigDecimal("12.21")));
+
+  private final Random random = new Random(871925);
+
+  @Test
+  public void testElementAccess() {
+    ValueArray arr = createArray(ELEMENTS);
+
+    assertThat(arr.numElements()).isEqualTo(3);
+    assertThat(arr.get(0)).isInstanceOf(VariantPrimitive.class);
+    assertThat(arr.get(0).asPrimitive().get()).isEqualTo(34);
+    assertThat(arr.get(1)).isInstanceOf(VariantPrimitive.class);
+    assertThat(arr.get(1).asPrimitive().get()).isEqualTo("iceberg");
+    assertThat(arr.get(2)).isInstanceOf(VariantPrimitive.class);
+    assertThat(arr.get(2).asPrimitive().get()).isEqualTo(new BigDecimal("12.21"));
+  }
+
+  @Test
+  public void testSerializationMinimalBuffer() {
+    ValueArray arr = createArray(ELEMENTS);
+
+    VariantValue value = roundTripMinimalBuffer(arr);
+
+    assertThat(value).isInstanceOf(SerializedArray.class);
+    SerializedArray actual = (SerializedArray) value;
+
+    assertThat(actual.numElements()).isEqualTo(3);
+    assertThat(actual.get(0)).isInstanceOf(VariantPrimitive.class);
+    assertThat(actual.get(0).asPrimitive().get()).isEqualTo(34);
+    assertThat(actual.get(1)).isInstanceOf(VariantPrimitive.class);
+    assertThat(actual.get(1).asPrimitive().get()).isEqualTo("iceberg");
+    assertThat(actual.get(2)).isInstanceOf(VariantPrimitive.class);
+    assertThat(actual.get(2).asPrimitive().get()).isEqualTo(new BigDecimal("12.21"));
+  }
+
+  @Test
+  public void testSerializationLargeBuffer() {
+    ValueArray arr = createArray(ELEMENTS);
+
+    VariantValue value = roundTripLargeBuffer(arr);
+
+    assertThat(value).isInstanceOf(SerializedArray.class);
+    SerializedArray actual = (SerializedArray) value;
+
+    assertThat(actual.numElements()).isEqualTo(3);
+    assertThat(actual.get(0)).isInstanceOf(VariantPrimitive.class);
+    assertThat(actual.get(0).asPrimitive().get()).isEqualTo(34);
+    assertThat(actual.get(1)).isInstanceOf(VariantPrimitive.class);
+    assertThat(actual.get(1).asPrimitive().get()).isEqualTo("iceberg");
+    assertThat(actual.get(2)).isInstanceOf(VariantPrimitive.class);
+    assertThat(actual.get(2).asPrimitive().get()).isEqualTo(new BigDecimal("12.21"));
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {300, 70_000, 16_777_300})
+  public void testMultiByteOffsets(int len) {
+    // Use a string exceeding 255 bytes to test value offset sizes of 2, 3, and 4 bytes
+    String randomString = RandomUtil.generateString(len, random);
+    VariantPrimitive bigString = Variants.of(randomString);
```

Comment:
Please fix warnings for things like this. `Raw use of parameterized class 'VariantPrimitive'`